safeweb/http.go

```diff
@@ -0,0 +1,255 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+// Package safeweb provides a simple, safe, and opinionated wrapper around
+// http.Server that applies some web application security defenses by default.
+// It is intended to be used in place of http.Server internal applications that
+// reside on the Tailscale network.
+//
+// safeweb requires that applications be built using two distinct
+// `http.ServeMux`s for serving browser and API clients to serve the appropriate
+// security defenses in their respective contexts. When serving requests it
+// will first attempt to route requests to the browser mux, and if no matching
+// route is found it will instead use the API mux.
+//
+// safeweb enforces Cross-Site Request Forgery (CSRF) protection for all routes
+// in the browser mux using the gorilla/csrf package. It is necessary to
+// template the CSRF token into all forms that are submitted to the server using
+// the `csrf.TemplateField` and `csrf.TemplateTag(r *http.Request)` APIs.
+//
+// safeweb expects either a Hostname or a Listener to be provided in the
+// configuration. If a hostname is provided it will create a tsnet server by
```

Comment:
Can we separate the tsnet concern out? It doesn't really seem necessary to have this library manage that plumbing directly, since we already provided a way to plumb in the network listener. 

At the risk of harping on my previous comment, I'd advocate for keeping this focused as narrowly as we can on HTTP servers, whether tsnet-based or otherwise. (We should of course make sure to compose well with tsnet, since we use it all over the place).

Reply:
I guess if we want to be able to set up redirectors, we might need to make the listener interface a little more general—say, perhaps, the caller gives us a function to call to _obtain_ listeners so we can listen on multiple ports?